packages/plugin-deduplicate/sources/commands/deduplicate.ts

```diff
@@ -0,0 +1,78 @@
+import {BaseCommand}                                    from '@berry/cli';
+import {Configuration, LocatorHash, Project, IdentHash} from '@berry/core';
+import {structUtils}                                    from '@berry/core';
+import {Command}                                        from 'clipanion';
+import * as semver                                      from 'semver';
+
+
+// eslint-disable-next-line arca/no-default-export
+export default class DeduplicateCommand extends BaseCommand {
+  static usage = Command.Usage({
+    category: `Workspace-related commands`,
+    description: `Reduces dependencies with overlapping ranges to a minimal set of packages`,
+    details: `https://github.com/atlassian/yarn-deduplicate for yarn v2`,
+    examples: [],
+  });
+
+  @Command.Path(`deduplicate`)
+  async execute() {
+    const configuration = await Configuration.find(this.context.cwd, this.context.plugins);
+    const {project} = await Project.find(configuration, this.context.cwd);
+
+    const locatorsByIdent: Map<IdentHash, Set<LocatorHash>> = new Map();
+    for (const [descriptorHash, locatorHash] of project.storedResolutions.entries()) {
+      const value = locatorHash;
+      const descriptor = project.storedDescriptors.get(descriptorHash)!
+      const key = descriptor.identHash;
+
+      const descriptors = locatorsByIdent.get(key);
+      if (descriptors === undefined) {
+        locatorsByIdent.set(key, new Set([value]));
+      } else {
+        locatorsByIdent.set(key, descriptors.add(value));
+      }
+    }
+
+    for (const descriptorHash of project.storedResolutions.keys()) {
```

Comment:
There's one subtlety: you're currently also iterating on the descriptors that are specifically mapped to a specific version via the `resolutions` field. I think to solve that you'd need to check whether the descriptor is covered by a resolution rule.

You can check `AliasResolver.resolve` to see how we do it - you currently have to copy/paste this logic, but I think it would make sense to move it inside a static method of `AliasResolver` so that consumers wouldn't have to reimplement it (since it's a bit tricky).

Reply:
Good point. Would this issue be "resolved" anyway since I have to run `yarn install` after this step to update the `.pnp` file? It wouldn't be a nice solution since dry-run would report a possible deduplication when it's explicitly broken by resolutions.

That's kind of why this isn't intended as a core plugin. Currently I'm using a simple "choose highest" strategy which isn't necessarily optimal. I would have to refresh my graph knowledge to understand what other strategies are possible and watch runtime (though the plugin is targeted at already almost minimal lockfiles so you wouldn't have more than a couple of duplicate packages before checking if there's a smaller cover)

Reply:
> Would this issue be "resolved" anyway since I have to run `yarn install` after this step to update the `.pnp` file?

I think it would generally be better, for the reason you mentioned and also because it could go the other way - in theory you could cause a package Foo@^1.0 to be affected by the enforced resolution of a package Foo@^1.5 (particularly if Foo@^1.5 is remapped to a different package, for example Bar@1.5). Big ball of hair 😅



> Currently I'm using a simple "choose highest" strategy which isn't necessarily optimal.

I was thinking about it - if we assume that a dependency tree only ever use caret `^` ranges, it's not so bad. All ranges will converge on using the highest caret-compatible semver range.

It's only when exact and tilde range enter the dance that you'll lose potentially better options, which might be reasonable enough ...

Reply:
> I was thinking about it - if we assume that a dependency tree only ever use caret ^ ranges, it's not so bad.

Yep. Thought about it a bit more and the more general case is that the semver ranges of the listed packages is equal. So either all use caret `^` or all use tilde `~`.